maven-core/src/main/java/org/apache/maven/execution/DefaultBuildResumptionManager.java

```diff
@@ -0,0 +1,308 @@
+     */
+    private Optional<String> getResumeFrom( MavenExecutionResult result, MavenProject failedProject )
+    {
+        List<MavenProject> allSortedProjects = result.getTopologicallySortedProjects();
+        if ( !allSortedProjects.get( 0 ).equals( failedProject ) )
+        {
+            return Optional.of( String.format( "%s:%s", failedProject.getGroupId(), failedProject.getArtifactId() ) );
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Projects after the first failed project could have succeeded by using -T or --fail-at-end.
+     * These projects can be skipped from later builds.
+     * This is not the case these projects are dependent on one of the failed projects.
+     * @param result The result of the Maven build.
+     * @param failedProjects The list of failed projects in the build.
+     * @param resumeFromProject The project where the build will be resumed with in the next run.
+     * @return An optional containing a comma separated list of projects which can be skipped,
+     *   or an empty optional if no projects can be skipped.
+     */
+    private Optional<String> determineProjectsToSkip( MavenExecutionResult result, List<MavenProject> failedProjects,
+                                                      MavenProject resumeFromProject )
+    {
+        List<MavenProject> allProjects = result.getTopologicallySortedProjects();
+        int resumeFromProjectIndex = allProjects.indexOf( resumeFromProject );
+        List<MavenProject> remainingProjects = allProjects.subList( resumeFromProjectIndex + 1, allProjects.size() );
+
+        List<GroupArtifactPair> failedProjectsGAList = failedProjects.stream()
+                .map( GroupArtifactPair::new )
+                .collect( Collectors.toList() );
+
+        String projectsToSkip = remainingProjects.stream()
+                .filter( project -> result.getBuildSummary( project ) instanceof BuildSuccess )
+                .filter( project -> hasNoDependencyOnProjects( project, failedProjectsGAList ) )
+                .map( project -> String.format( "%s:%s", project.getGroupId(), project.getArtifactId() ) )
+                .collect( Collectors.joining( PROPERTY_DELIMITER ) );
+
+        if ( !StringUtils.isEmpty( projectsToSkip ) )
+        {
+            return Optional.of( projectsToSkip );
+        }
+
+        return Optional.empty();
+    }
+
+    private boolean hasNoDependencyOnProjects( MavenProject project, List<GroupArtifactPair> projectsGAs )
+    {
+        return project.getDependencies().stream()
+                .map( GroupArtifactPair::new )
+                .noneMatch( projectsGAs::contains );
+    }
+
+    private boolean writeResumptionFile( MavenProject rootProject, Properties properties )
+    {
+        Path resumeProperties = Paths.get( rootProject.getBuild().getDirectory(), RESUME_PROPERTIES_FILENAME );
+        try
+        {
+            Files.createDirectories( resumeProperties.getParent() );
```

Comment:
Does this one assume that we are in a phase/goal where Maven hasn't created `target` yet?

Reply:
Yes, in the case of our [integration test project](https://github.com/apache/maven-integration-testing/pull/61/files), we execute `mvn test` on a multi-module project. (root + 3 children). The root project will not have a target folder generated at the end of the build.